streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/ClockService.scala

```diff
@@ -75,10 +82,12 @@ class ClockService(dag : DAG, store: AppDataStore) extends Actor with Stash {
       processorIdWithDescription =>
         val (processorId, description) = processorIdWithDescription
         val taskClocks = new Array[TimeStamp](description.parallelism).map(_ => startClock)
-        val processorClock = new ProcessorClock(processorId, startClock, taskClocks)
-        processorClocks.add(processorClock)
+        val processorClock = new ProcessorClock(taskClocks)
         processorClockLookup.put(processorId, processorClock)
     }
+
+    this.processorIdToLevel = dag.graph.topologicalOrderIterator.zipWithIndex.toMap
```

Comment:
Why this virtual link is needed? Is there any better way to make topo sort stable? E.g. when there are multiple nodes without input, additionally sort them with nodeId. Something like that.